test/arrays.jl

```diff
@@ -0,0 +1,504 @@
+
+    # works since aliasing is "simple"
+    autodiff(Reverse, foo_bc!, Const, Duplicated(R, dR), Duplicated(R, dR), Duplicated(B, dB))
+
+    A = rand(10,10); B = rand(10, 10)
+    dA = zero(A); dB = zero(B); dR = fill!(similar(A), 1)
+
+    autodiff(Reverse, foo_bc!, Const, Duplicated(A, dR), Duplicated(transpose(A), transpose(dA)), Duplicated(B, dB))
+end
+
+@testset "Array Copy" begin
+	F = [2.0, 3.0]
+
+	dF = [0.0, 0.0]
+
+	function copytest(F)
+		F2 = copy(F)
+		@inbounds F[1] = 1.234
+		@inbounds F[2] = 5.678
+		@inbounds F2[1] * F2[2]
+	end
+	autodiff(Reverse, copytest, Duplicated(F, dF))
+	@test F ≈ [1.234, 5.678]
+	@test dF ≈ [3.0, 2.0]
+
+    @test 31.0 ≈ autodiff(Forward, copytest, Duplicated([2.0, 3.0], [7.0, 5.0]))[1]
+
+    function sh(x)
+        Base.sizehint!(x, length(x))
+        nothing
+    end
+
+    autodiff(Reverse, sh, Duplicated([1.0], [0.0]))
+end
+
+@testset "Arrays are double pointers" begin
+    @noinline function func_scalar(X)
+        return X
+    end
+
+    function timsteploop_scalar(FH1)
+        G = Float64[FH1]
+        k1 = @inbounds func_scalar(G[1])
+        return k1
+    end
+    @test Enzyme.autodiff(Reverse, timsteploop_scalar, Active(2.0))[1][1] ≈ 1.0
+    @test Enzyme.autodiff(Forward, timsteploop_scalar, Duplicated(2.0, 1.0))[1] ≈ 1.0
+
+    @noinline function func(X)
+        return @inbounds X[1]
+    end
+    function timsteploop(FH1)
+        G = Float64[FH1]
+        k1 = func(G)
+        return k1
+    end
+    @test Enzyme.autodiff(Reverse, timsteploop, Active(2.0))[1][1] ≈ 1.0
+    @test Enzyme.autodiff(Forward, timsteploop, Duplicated(2.0, 1.0))[1] ≈ 1.0
+end
+
+function bc0_test_function(ps)
+    z = view(ps, 26:30)
+    C = Matrix{Float64}(undef, 5, 1)
+    C .= z
+    return C[1]
+end
+
+@noinline function bc1_bcs2(x, y)
+    x != y && error(2)
+    return x
+end
+
+@noinline function bc1_affine_normalize(x::AbstractArray)
+    _axes = bc1_bcs2(axes(x), axes(x))
+    dest = similar(Array{Float32}, _axes)
+    bc = convert(Broadcast.Broadcasted{Nothing}, Broadcast.instantiate(Base.broadcasted(+, x, x)))
+    copyto!(dest, bc)
+    return x
+end
+
+function bc1_loss_function(x)
+    return bc1_affine_normalize(x)[1]
+end
+
+function bc2_affine_normalize(::typeof(identity), x::AbstractArray, xmean, xvar,
+    scale::AbstractArray, bias::AbstractArray, epsilon::Real)
+    _scale = @. scale / sqrt(xvar + epsilon)
+    _bias = @. bias - xmean * _scale
+    return @. x * _scale + _bias
+end
+
+function bc2_loss_function(x, scale, bias)
+    x_ = reshape(x, 6, 6, 3, 2, 2)
+    scale_ = reshape(scale, 1, 1, 3, 2, 1)
+    bias_ = reshape(bias, 1, 1, 3, 2, 1)
+
+    xmean = mean(x_, dims=(1, 2, 5))
+    xvar = var(x_, corrected=false, mean=xmean, dims=(1, 2, 5))
+
+    return sum(abs2, bc2_affine_normalize(identity, x_, xmean, xvar, scale_, bias_, 1e-5))
+end
+
+#WARN: why is this segfaulting right now???
+#=
+@testset "Broadcast noalias" begin
```

Comment:
@wsmoses this is mysteriously segfaulting for me now.  Not sure if I broke this or this is the same segfault we've seen in other tests?

Reply:
uh it should not be segfaulting and isn't presently